plugins/nf-nomad/src/main/nextflow/nomad/executor/NomadService.groovy

```diff
@@ -18,93 +18,136 @@
- *
  * @author Abhinav Sharma <abhi18av@outlook.com>
  */
 
 @Slf4j
 @CompileStatic
-class NomadService implements Closeable {
+class NomadService implements Closeable{
 
-    NomadConfig config
+    private final nextflow.nomad.NomadConfig config
 
+    private final JobsApi jobsApi
 
-    NomadService(NomadExecutor executor) {
-        assert executor
-        this.config = executor.config
+    NomadService(nextflow.nomad.NomadConfig config) {
+        this.config = config
+        ApiClient apiClient = new ApiClient()
+        apiClient.basePath = config.clientOpts.address
+        this.jobsApi = new JobsApi(apiClient);
     }
 
+    @Override
+    void close() throws IOException {
+    }
 
-    @Memoized
-    protected NomadClient getClient() {
-        new NomadClient(config)
+    String submitTask(String id, String name, String image,
+                      List<String> args,
+                      String workingDir,
+                      Map<String, String>env){
+        Job job = new Job();
+        job.ID = id
+        job.name = name
+        job.type = "batch"
+        job.datacenters = this.config.jobOpts.datacenters
+        job.namespace = this.config.jobOpts.namespace
+
+        job.taskGroups = [createTaskGroup(id, name, image, args, workingDir, env)]
+
+        JobRegisterRequest jobRegisterRequest = new JobRegisterRequest();
+        jobRegisterRequest.setJob(job);
+        JobRegisterResponse jobRegisterResponse = jobsApi.registerJob(jobRegisterRequest, config.jobOpts.region, config.jobOpts.namespace, null, null)
+        jobRegisterResponse.evalID
     }
 
-    @Override
-    void close() {
-        Global.onCleanup((it) -> client.closeConnection())
+    TaskGroup createTaskGroup(String id, String name, String image, List<String> args, String workingDir, Map<String, String>env){
+        def task = createTask(id, image, args, workingDir, env)
+        def taskGroup = new TaskGroup(
+                name: "group",
+                tasks: [ task ]
+        )
+        return taskGroup
+    }
 
+    Task createTask(String id, String image, List<String> args, String workingDir, Map<String, String>env) {
+        def task = new Task(
+                name: "nf-task",
+                driver: "docker",
+                config: [
+                        image: image,
+                        privileged: true,
+                        work_dir: workingDir,
+                        command: args.first(),
+                        args: args.tail(),
+                ] as Map<String,Object>,
+                env: env
+        )
+        if( config.jobOpts.dockerVolume){
+            String destinationDir = workingDir.split(File.separator).dropRight(2).join(File.separator)
+            task.config.mount = [
+                    type : "volume",
+                    target : destinationDir,
+                    source : config.jobOpts.dockerVolume,
+                    readonly : false
+            ]
+        }
```

Comment:
while docker volumes is a quick way to prototype adding storage to a job/task it not considered best practice. Since it is very driver specific and bypasses any native storage handling in Nomad. Also most production clusters will lock down mounting volumes this way for security reasons. So we should find a user friendly way to do this using the nomad `volume` block.

Reply:
Yes, I agree

by the moment the alternative/workaround is to use an s3 + fusion (and let `dockerVolume` empty)